src/System.Linq.Parallel/tests/ExchangeTests.cs

```diff
@@ -10,183 +11,177 @@ namespace Test
+        /// These ODEs either should not occur (preferred), or should not get into the aggregate exception.
+        ///
+        /// Also applies to the standard stand-alone chunk partitioner.
+        /// Does not apply to other partitioners unless an exception in one consumer would cause flow-on exception in others.
         /// </summary>
-        private static void OrderedPipeliningTest1(int dataSize, bool buffered)
+        /// <returns></returns>
+        [Theory]
+        [MemberData("BinaryRanges", new int[] { 16 }, new int[] { 16 }, MemberType = typeof(UnorderedSources))]
+        public static void PlinqChunkPartitioner_DontEnumerateAfterException(Labeled<ParallelQuery<int>> left, int leftCount, Labeled<ParallelQuery<int>> right, int rightCount)
         {
-            ParallelMergeOptions merge = buffered ? ParallelMergeOptions.FullyBuffered : ParallelMergeOptions.NotBuffered;
-
-            IEnumerable<int> src = Enumerable.Range(0, dataSize);
-            if (!Enumerable.SequenceEqual(src.AsParallel().AsOrdered().WithMergeOptions(merge).Select(x => x), src))
-            {
-                Assert.True(false, string.Format("OrderedPipeliningTest1: dataSize={0}, buffered={1}:  > FAILED: Incorrect output.", dataSize, buffered));
-            }
+            ParallelQuery<int> query =
+                left.Item.WithExecutionMode(ParallelExecutionMode.ForceParallelism)
+                    .Select(x => { if (x == 4) throw new DeliberateTestException(); return x; })
+                    .Zip(right.Item, (a, b) => a + b)
+                .AsParallel().WithExecutionMode(ParallelExecutionMode.ForceParallelism);
+
+            AggregateException ae = Assert.Throws<AggregateException>(() => query.ToArray());
+            Assert.Single(ae.InnerExceptions);
+            Assert.All(ae.InnerExceptions, e => Assert.IsType<DeliberateTestException>(e));
         }
 
         /// <summary>
-        /// Checks whether an ordered pipelining merge pipelines the results
-        /// instead of running in a stop-and-go fashion.
+        /// The stand-alone chunk partitioner calls takes an IEnumerator over the source, and disposes the enumerator when it is
+        /// finished.
+        /// If an exception occurs, the calling enumerator disposes the enumerator... but then other callers generate ODEs.
+        /// These ODEs either should not occur (preferred), or should not get into the aggregate exception.
+        ///
+        /// Also applies to the plinq stand-alone chunk partitioner.
+        /// Does not apply to other partitioners unless an exception in one consumer would cause flow-on exception in others.
         /// </summary>
-        private static void OrderedPipeliningTest2(bool buffered)
+        /// <returns></returns>
+        [Theory]
+        [MemberData("BinaryRanges", new int[] { 16 }, new int[] { 16 }, MemberType = typeof(UnorderedSources))]
```

Comment:
I'm having trouble following along with all of these theories, in particular because I have to jump somewhere else in the code to understand what function is feeding in the data, with arguments to that other function being provided in this attribute without names.  Can we do something to make this all more understandable, e.g. use comments consistently to describe what these things all are?